README.md
Visualization Quality Control
=============================

[![Join the chat at https://gitter.im/rmflight/visualizationQualityControl](https://badges.gitter.im/rmflight/visualizationQualityControl.svg)](https://gitter.im/rmflight/visualizationQualityControl?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Set of useful functions for calculating various measures from data and
visualizing them.

Takes a lot of inspiration from [Gierlinski et al.,
2015](https://dx.doi.org/10.1093/bioinformatics/btv425), especially the
`median_correlation` and `outlier_fraction` functions.

This **readme** is documenting functionality as of **818b62f8**.

Installation
------------

### Dependencies

Note that before installing, you will want to install the `ggbiplot`
package, and at least `v1.2.1` of the `ComplexHeatmap` package. Robert M
Flight maintains a fork of `ggbiplot` on GitHub because it is not part
of CRAN, and as of July 2, 2015, `ComplexHeatmap` must be installed from
GitHub:

    devtools::install_github("rmflight/ggbiplot")
    devtools::install_github("jokergoo/ComplexHeatmap")

Other odd dependencies that may not be present include the `dendsort`
package, and the `viridis` package:

    install.packages("dendsort")
    install.packages("viridis")

### This Package

This package can be installed by cloning from the GitLab repo:

    git clone https://gitlab.cesb.uky.edu/rmflight/visualizationQualityControl.git
    cd visualizationQualityControl
    R
    devtools::install(".", quick = FALSE) # builds the vignette, which you definitely want

Alternatively, you can install it from GitHub in one go:

    devtools::install_github("rmflight/visualizationQualityControl", quick = FALSE)

Examples
--------

These examples show the primary functionality. We will apply the
visualizations to a two group dataset. However, all of the functions are
still applicable to datasets with more than two groups. The examples
below are for a dataset where there has been a sample swapped between
the two groups (i.e. there is a problem!). If you want to see how the
visualizations compare between a **good** dataset and a **bad** dataset,
see the vignette.

    library(visualizationQualityControl)
    data("grp_cor_data")
    exp_data <- grp_cor_data$data
    rownames(exp_data) <- paste0("f", seq(1, nrow(exp_data)))
    colnames(exp_data) <- paste0("s", seq(1, ncol(exp_data)))

    sample_info <- data.frame(id = colnames(exp_data), class = grp_cor_data$class)

    exp_data[, 5] <- grp_cor_data$data[, 19]
    exp_data[, 19] <- grp_cor_data$data[, 5]
    sample_classes <- sample_info$class

### visqc\_pca

    pca_data <- prcomp(t(exp_data), center = TRUE)
    visqc_pca(pca_data, groups = sample_classes)

![](README_files/figure-markdown_strict/do_pca-1.png)<!-- -->

### visqc\_heatmap

Calculate sample-sample correlations and reorder based on within class
correlations

    data_cor <- pairwise_correlation(t(exp_data), exclude_0 = TRUE)$cor
    data_order <- similarity_reorderbyclass(data_cor, sample_classes, transform = "sub_1")

And then generate a colormapping for the sample classes and plot the
correlation heatmap.

    data_legend <- generate_group_colors(2)
    names(data_legend) <- c("grp1", "grp2")
    row_data <- sample_info[, "class", drop = FALSE]
    row_annotation <- list(class = data_legend)

    library(viridis)
    library(circlize)
    colormap <- colorRamp2(seq(0.4, 1, length.out = 20), viridis::viridis(20))

    visqc_heatmap(data_cor, colormap, "Correlation", row_color_data = row_data,
                  row_color_list = row_annotation, col_color_data = row_data,
                  col_color_list = row_annotation, row_order = data_order$indices,
                  column_order = data_order$indices)

![](README_files/figure-markdown_strict/colormapandheatmap-1.png)<!-- -->

### median\_correlations

    data_medcor <- median_correlations(data_cor, sample_classes)
    ggplot(data_medcor, aes(x = sample_id, y = med_cor)) + geom_point() + 
      facet_grid(. ~ sample_class, scales = "free") + ggtitle("Median Correlation")

![](README_files/figure-markdown_strict/median_correlations-1.png)<!-- -->

### outlier\_fraction

    data_outlier <- outlier_fraction(t(exp_data), sample_classes)
    ggplot(data_outlier, aes(x = sample, y = frac)) + geom_point() + 
      facet_grid(. ~ class, scales = "free") + ggtitle("Outlier Fraction")

![](README_files/figure-markdown_strict/outlier_fraction-1.png)<!-- -->

Open Vignette
-------------

To open the vignette giving an example of examining data for quality
control purposes, you should see the `quality_control` vignette using:

    vignette("quality_control", package = "visualizationQualityControl")

This will open the vignette in the help pane in `RStudio`, which is
often what you want to happen.

Fake Data Generation
--------------------

Some fake data is stored in `grp_cor_data` that is useful for testing
the `median_correlation` function. It was generated by:

    library(fakeDataWithError)
    set.seed(1234)

    s1 <- runif(100, 0, 1)
    grp1 <- add_uniform_noise(10, s1, 0.1)

    model_data <- data.frame(s1 = s1, s2 = grp1[, 1])

    lm_1 <- lm(s1 ~ s2, data = model_data)

    lm_1$coefficients[2] <- 0.5

    s3 <- predict(lm_1)
    s4 <- add_uniform_noise(1, s3, 0.2)

    grp2 <- add_uniform_noise(10, s4, 0.1)

    grp_class <- rep(c("grp1", "grp2"), each = 10)

    grp_cor_data <- list(data = cbind(grp1, grp2), class = grp_class)

    library(fakeDataWithError)
    set.seed(1234)

    n_point <- 1000
    n_rep <- 10

    # a nice log-normal distribution of points with points along the entire range
    simulated_data <- c(rlnorm(n_point / 2, meanlog = 1, sdlog = 1),
                        runif(n_point / 2, 5, 100))

    # go to log to have decent correlations on the "transformed" data
    lsim1 <- log(simulated_data)

    # add some uniform noise to get lower than 1 correlations
    lgrp1 <- add_uniform_noise(n_rep, lsim1, .5)

    # add some uniform noise to everything in normal space
    sim1_error <- add_uniform_noise(n_rep, simulated_data, 1, use_zero = TRUE)
    # and generate the grp1 data in normal space
    ngrp1 <- exp(lgrp1) + sim1_error


    # do regression to generate some other data
    model_data <- data.frame(lsim1 = lsim1, lsim2 = lgrp1[, 1])
    lm_1 <- lm(lsim1 ~ lsim2, data = model_data)

    # reduce the correlation between them
    lm_1$coefficients[2] <- 0.5
    lsim3 <- predict(lm_1)

    # and a bunch of error
    lsim4 <- add_uniform_noise(1, lsim3, 1.5)

    # create group with added error to reduce correlation from 1
    lgrp2 <- add_uniform_noise(10, lsim4, .5)

    # add error in original space
    nsim4 <- exp(lsim4)
    sim4_error <- add_uniform_noise(10, nsim4, 1, use_zero = TRUE)
    ngrp2 <- exp(lgrp2) + sim4_error

    # put all data together, and make negatives zero
    all_data <- cbind(ngrp1, ngrp2)
    all_data[(all_data < 0)] <- 0

    grp_class <- rep(c("grp1", "grp2"), each = 10)

    grp_exp_data <- list(data = all_data, class = grp_class)


NEWS.md
# 0.3.0

* `median_correlations` gains a new argument, `between_classes` to generate the
median values to samples in other classes. This causes the appearance of two
more columns when set to TRUE. The default is FALSE, so hopefully this does not
cause current code to misbehave, but I've bumped the version number as a warning.

# v 0.2.18

* Augmented correlations (`weight = TRUE`) should be much more useful and interpretable.

* `information_volume` and `correspondence` calculations improved. Namely that
`information_volume` is being scaled by the maximum. 

* `correspondence` by default **does not** consider presence of zeros in both
samples to be informative, this can be changed by setting `not_both = TRUE`. The
default is more useful in cases where there are lots of features and the data is
sparse, and zeros are likely to happen by chance.

* In addition to returning the `cor` matrix and `keep` matrix, `pairwise_correlations`
now returns the `raw` correlations, and the weighting matrices `info` and `correspondence`
so that each one can be examined.

* The diagonal of `info` weighting corresponds to how many features a sample has
compared to the sample with the most features.

# v 0.2.5

* Added two functions, `information_volume` and `correspondence` to calculate
weights based on the amount of things that are non-zero in both things when
doing pairwise correlation.

* Added logical argument `weight` to `pairwise_correlation` to weight the correlations. If `weight = TRUE`, the diagonal will not be **1** anymore, but instead will reflect how many features out of the total are in that sample.

# v 0.2.3

* A bug was discovered in `median_correlations` that meant the wrong sample ids
might be added to the output data, making detection of real problems difficult

# v 0.2.1

* `pairwise_correlation` now uses `cor` internally directly, whereas previously
it did a `for` loop to allow pairwise comparisons. This makes the correlations
3x faster.

* `count` has been removed from the list returned by `pairwise_correlation`

* new function `pairwise_correlation_count` to get the counts in each pairwise
comparison
 
# v 0.1.1

* Changed correlation function to return a list instead of a matrix. This
list contains the correlations (`cor`), counts in each correlation (`count`),
and which points passed the criteria (`keep`).


tests/testthat.R
library(testthat)
library(visualizationQualityControl)

test_check("visualizationQualityControl")


tests/testthat/test_filtering.R
context("data filtering")

# set up the data
test_data <- matrix(rnorm(80, mean = 20), nrow = 10, ncol = 8)
rownames(test_data) <- as.character(seq(1, nrow(test_data)))
test_data[1, c(1, 3, 6)] <- 0
test_data[3, c(1,2,3)] <- 0
test_data[4, c(1,2,3,4)] <- 0

test_data[7, 7] <- 0
test_data[8, c(1, 2, 6, 8)] <- 0

sample_classes <- rep(c("A", "B"), each = 4)


test_that("integer filtering works without classes", {
          f_1 <- t(keep_non_zero_percentage(t(test_data), keep_num = 5))
          expect_equal(c("1", "2", "3", "5", "6", "7", "9", "10"), rownames(f_1))
          })
  

test_that("integer filtering works with classes", {
          f_2 <- t(keep_non_zero_percentage(t(test_data), sample_classes, 3))
          expect_equal(c("1", "2", "3", "4", "5", "6", "7", "9", "10"), rownames(f_2))
          })

test_that("percentage filtering works without classes", {
          f_3 <- t(keep_non_zero_percentage(t(test_data), keep_num = 0.6))
          expect_equal(c("1", "2", "3", "5", "6", "7", "9", "10"), rownames(f_3))
          })

test_that("percentage filtering works with classes", {
          f_4 <- t(keep_non_zero_percentage(t(test_data), sample_classes, 0.7))
          expect_equal(c("1", "2", "3", "4", "5", "6", "7", "9", "10"), rownames(f_4))
          })

test_that("error handling works", {
  expect_error(keep_non_zero_percentage(t(test_data), sample_classes, -1))
  expect_error(keep_non_zero_percentage(t(test_data), sample_classes, 100))
})


tests/testthat/test-classcolors.R
context("classcolors")


test_that("defaults work", {
  col_1 <- generate_group_colors(3)
  col_2 <- generate_group_colors(3)
  
  expect_equal(col_1, col_2)
  
  col_3 <- generate_group_colors(5)
  col_4 <- generate_group_colors(5)
  
  expect_true(!(identical(col_3, col_4)))
})

test_that("overrides work", {
  col_1 <- generate_group_colors(3)
  col_2 <- generate_group_colors(3, TRUE)
  expect_true(!(identical(col_1, col_2)))
  
  col_3 <- generate_group_colors(5, FALSE)
  col_4 <- generate_group_colors(5, FALSE)
  expect_equal(col_3, col_4)
  
  col_5 <- generate_group_colors(5, TRUE)
  expect_true(!(identical(col_3, col_5)))
})

test_that("setting set.seed works", {
  set.seed(1234)
  col_1 <- generate_group_colors(5)
  
  set.seed(1234)
  col_2 <- generate_group_colors(5, TRUE)
  expect_equal(col_1, col_2)
})


README.R
## ----load_data-----------------------------------------------------------
library(visualizationQualityControl)
data(all_intensity)
data(all_info)

## ----dopca---------------------------------------------------------------
int_zero <- all_intensity
int_zero[is.na(int_zero)] <- 0
has_3 <- apply(int_zero, 1, function(x){sum(x != 0) >= 2})
int_zero <- int_zero[has_3,]
all_pca <- prcomp(t(int_zero), center = TRUE, scale. = TRUE)

## ----plotpca-------------------------------------------------------------
visqc_pca(all_pca, groups = all_info$disease)

## ----calc_correlation----------------------------------------------------
all_cor <- pairwise_correlation(t(all_intensity))
dim(all_cor)

## ----plot_heatmap--------------------------------------------------------
library(circlize)
colormap <- colorRamp2(c(0.95, 1), c("black", "white"))
visqc_heatmap(all_cor, colormap)

## ----reorder-------------------------------------------------------------
all_order <- similarity_reorderbyclass(all_cor, all_info[, "disease", drop = FALSE], transform = "sub_1")
visqc_heatmap(all_cor, colormap, row_order = all_order$indices, column_order = all_order$indices)

## ----add_legend----------------------------------------------------------
color_legend <- generate_group_colors(2)
names(color_legend) <- c("normal", "cancer")

row_data <- all_info[, "disease", drop = FALSE]
row_annotation <- list(disease = color_legend)
visqc_heatmap(all_cor, colormap, row_color_data = row_data, row_color_list = row_annotation, col_color_data = row_data, col_color_list = row_annotation, row_order = all_order$indices, column_order = all_order$indices)

## ----add_legend2---------------------------------------------------------
all_info$random
color_legend <- generate_group_colors(6)
names(color_legend) <- c(unique(all_info$random), unique(all_info$disease))

col_data <- all_info
col_annotation <- list(random = color_legend[1:4], disease = color_legend[5:6])

row_data <- all_info[, "disease", drop = FALSE]
row_annotation <- list(disease = color_legend[5:6])
visqc_heatmap(all_cor, colormap, row_color_data = row_data, row_color_list = row_annotation, col_color_data = col_data, col_color_list = col_annotation, row_order = all_order$indices, column_order = all_order$indices)



R/data_filtering.R
#' keep features with percentage of non-zeros
#' 
#' Given a value matrix (features are columns, samples are rows), and sample classes, 
#' find those things that are not zero in at least a certain 
#' number of one of the classes, and keep them
#' 
#' @param data_matrix the matrix of values to work with 
#' @param sample_classes the classes of each sample
#' @param keep_num what number of samples in each class need a non-zero value (see Details)
#' 
#' @details This function is being deprecated and all code should use the
#'   \code{keep_non_zero_percentage} function instead.
#' 
#' @seealso keep_non_zero_percentage
#' @export
#' @return matrix
filter_non_zero_percentage <- function(data_matrix, sample_classes = NULL, keep_num = 0.75){
  .Deprecated("keep_non_zero_percentage", "visualizationQualityControl")
  keep_non_zero_percentage(data_matrix, sample_classes, keep_num)
}

#' keep features with percentage of non-zeros
#' 
#' Given a value matrix (features are columns, samples are rows), and sample classes, 
#' find those things that are \emph{not zero} in at least a certain 
#' number of samples in one of the classes, and keep those features for further
#' processing.
#' 
#' @param data_matrix the matrix of values to work with 
#' @param sample_classes the classes of each sample
#' @param keep_num what number of samples in each class need a non-zero value (see Details)
#' @param zero_value what number represents zero values
#' 
#' @details The number of samples that must be non-zero can be expressed either as a whole
#'   number (that is greater than one), or as a fraction that will be be multiplied 
#'   by the number of samples in each class to get the lower limits for each of the classes.
#' 
#' @export
#' @return matrix
keep_non_zero_percentage <- function(data_matrix, sample_classes = NULL, keep_num = 0.75, zero_value = 0){
  stopifnot(ncol(data_matrix) != 0)
  stopifnot(nrow(data_matrix) != 0)
  stopifnot(keep_num >= 0)
  stopifnot(!(keep_num >= nrow(data_matrix)))
  
  if (is.null(sample_classes)) {
    sample_classes <- rep("A", nrow(data_matrix))
  }
  uniq_classes <- unique(sample_classes)
  class_index <- lapply(uniq_classes, function(x){sample_classes %in% x})
  names(class_index) <- uniq_classes
  
  if (keep_num < 1) {
    min_notzero <- sapply(class_index, function(x){round(sum(x) * keep_num)})
  } else {
    min_notzero <- sapply(class_index, function(x){round(keep_num)})
  }
  
  # what is this doing?
  # For each of the features (columns), check how many non-zero entries
  # there are for each class. If the number is greater than the limit
  # in at least one of the classes, keep that feature. This allows easy filtering
  # of those features that have more than the specified number of zeros in all
  # classes.
  # 
  # iterate over columns (features)
  has_min <- apply(data_matrix, 2, function(in_col){
    # how many are non-zero in each class
    n_pass <- sapply(class_index, function(index){sum(in_col[index] > zero_value)})
    # is the minimum reached in at least one class
    keep <- sum(n_pass >= min_notzero) > 0
    keep
  })
  data_matrix[, has_min]
}


R/sd_rsd_summaries.R
#' summarize data
#'
#' summarizes a matrix or data.frame, where rows are samples
#' and columns are features
#'
#' @param in_data matrix or data.frame
#' @param sample_classes which samples are in which class
#' @param avg_function which function to use for summary
#' @param log_transform apply a log-transform to the mean
#' @param remove_zeros remove zeros before summarizing
#'
#' @return data.frame
#' @export
summarize_data <- function(in_data, sample_classes=NULL, avg_function = mean,
                           log_transform = FALSE, remove_zeros = FALSE){
  if (is.null(sample_classes)){
    sample_classes <- rep("A", nrow(in_data))
  }

  n_feature <- ncol(in_data)
  split_indices <- split(seq(1, nrow(in_data)), sample_classes)

  split_values <- lapply(split_indices, function(use_index){
    tmp_mean_sd <- apply(in_data[use_index, , drop = FALSE], 2, function(x){
      if (remove_zeros) {
        x <- x[x != 0]
      }
      c(avg_function(x), sd(x), max(x) - min(x))
    })
    tmp_mean_sd <- t(tmp_mean_sd)
    colnames(tmp_mean_sd) <- c("mean", "sd", "diff")
    data.frame(mean = tmp_mean_sd[, "mean"],
               var = c(tmp_mean_sd[, "sd"],
                       tmp_mean_sd[, "sd"] / tmp_mean_sd[, "mean"],
                       tmp_mean_sd[, "diff"]),
               type = rep(c("sd", "rsd", "diff"), each = n_feature))
  })

  out_data <- do.call(rbind, split_values)
  class_rep <- vapply(split_values, nrow, numeric(1))
  out_data$class <- rep(names(split_values), times = class_rep)
  out_data$type <- factor(out_data$type, ordered = TRUE, levels = c("diff", "sd", "rsd"))

  if (is.function(log_transform)) {
    out_data$log_mean <- log_transform(out_data$mean)
  }
  out_data
}




#' calculate F-ratio
#'
#' given a data matrix of samples (rows) and features (columns), and a vector of classes (character or factor),
#' calculate an F-ratio for each feature.
#'
#' @param data the data matrix, with samples (rows) and features (columns)
#' @param data_classes what are the classes of the rows
#'
#' @return vector
#' @export
calculate_fratio <- function(data, data_classes){
  if(is.character(data_classes)){
    data_classes <- factor(data_classes)
  }

  all_means <- colMeans(data)

  split_indices <- split(seq(1, nrow(data)), data_classes)
  n_sample <- nrow(data)
  n_group <- length(split_indices)
  split_data <- lapply(split_indices, function(in_index){data[in_index, , drop = FALSE]})

  group_means <- lapply(split_data, colMeans)
  group_var <- lapply(split_data, function(in_data){apply(in_data, 2, var)})
  group_count <- lapply(split_data, nrow)

  weight_var <- function(count, var, sub1 = TRUE){
    if (sub1){
      count <- count - 1
    }
    count * var
  }
  within_var <- Map(weight_var, group_count, group_var)
  within_var <- do.call(rbind, within_var)
  within_var <- colSums(within_var) / (n_sample - n_group)

  between_var <- lapply(group_means, function(in_mean){(in_mean - all_means)^2})
  between_var <- Map(weight_var, group_count, between_var, FALSE)
  between_var <- do.call(rbind, between_var)
  between_var <- colSums(between_var) / (n_group - 1)

  f_ratio <- between_var / within_var
  f_ratio
}

#' calculate values from summaries
#'
#' given a data.frame of means and variances, calculate mean sd at low end and
#' mean rsd at high end.
#'
#' @param data data.frame of means and variances
#' @param low_cut means <= this value used for average sd
#' @param hi_cut means >= this value used for average rsd
#'
#' @import dplyr
#' @export
#' @return vector
calc_sd_rsd <- function(data, low_cut, hi_cut = NULL){
  if (is.null(hi_cut)){
    hi_cut <- min(data[, "mean"]) > low_cut
  }

  sd_mn <- filter(data, mean <= low_cut, type == "sd") %>% summarise(., mean(var))
  rsd_mn <- filter(data, mean >= hi_cut, type == "rsd") %>% summarise(., mean(var))

  return(c(sd = sd_mn, rsd = rsd_mn))
}

#' calculate values from summaries v2
#'
#' given a data.frame of means and variances, use a two step non-linear least squares.
#' The first step is done on the mean vs sd, then the estimates are used in a second
#' that estimates them using the mean vs rsd.
#'
#' @param data data.frame of means and variances
#' @param ... other nls parameters
#'
#' @import dplyr
#' @export
#' @return vector
calc_sd_rsd_nls <- function(data, ...){
  nl_sd <- filter(data, type == "sd") %>% nls(var ~ B + A*mean, data = ., start = list(A = 0, B = 0), ...)
  nl_rsd <- filter(data, type == "rsd") %>% nls(var ~ ((A * mean) + B) / mean, data = ., start = list(A = coef(nl_sd)["A"], B = coef(nl_sd)["B"]), ...)

  return(c(additive = coef(nl_rsd)["B"], proportional = coef(nl_rsd)["A"]))
}


R/principal_components.R
#' sensible principal components
#' 
#' given the results of \code{\link{prcomp}}, generate decent looking 
#' plot of selected components using \code{ggbiplot}.
#' 
#' @param pca_decomp the results of \code{prcomp}
#' @param princomps which principal components to plot
#' @param groups factor defining the classes of each object
#' @param labels labels of the points
#' @param dot_size size of the points (default is 1)
#'
#' @import ggbiplot
#' @import ggplot2
#' @export
visqc_pca <- function(pca_decomp, princomps = c(1, 2), groups = NULL, labels = NULL, dot_size = 4){
  x_data <- pca_decomp$x
  n_obj <- nrow(x_data)
  
  if (is.null(groups)){
    groups <- as.factor(rep("", n_obj))
  } else if (is.data.frame(groups)){
    groups <- as.factor(apply(groups, 1, paste, collapse = "."))
  }
  
  
  ggbiplot(pca_decomp, obs.scale = 1, groups = groups, labels = labels, var.axes = FALSE, choices = princomps) + geom_point(aes(color = groups), size = dot_size) + coord_cartesian()
}


R/heatmaps.R
#' 10 sample to sample correlations
#' 
#' test data of 10 sample to sample correlations where samples are drawn
#' from two groups. Generated by rmflight from random distributions
#' 
#' @format matrix with 10 rows and 10 columns, with row and colnames
#' @source generated by rmflight
#' @name grp_cor
NULL

#' 10 sample meta-data
#' 
#' meta-data for \code{grp_cor}. 
#' 
#' @format data.frame with 10 rows, and 2 columns, \code{grp} defining with group
#' and \code{set}, defining the set.
#' @source generated by rmflight
#' @name grp_info
NULL

#' create set of disjoint colors
#' 
#' When multiple sample classes need to be visualized on a heatmap, it is useful
#' to be able to distinguish them by color. This function generates a set of colors
#' for sample classes
#' 
#' @param n_group how many groups should there be colors for
#' @param randomize should colors be randomized? (default is \code{NULL}). See \emph{details}.
#' 
#' @details the default for \code{randomize} is \code{NULL}, so that reordering
#'   the colors randomly is decided purely based on the number of colors requested.
#'   Currently, that cutoff is \emph{5} colors, less than that the colors will
#'   always be in the same order, for \emph{5} colors or more, they will
#'   be in a scrambled order, different each time unless \code{set.seed} is
#'   used. If \code{randomize} is \code{TRUE} or \code{FALSE}, then it overrides
#'   the defaults.
#' 
#' @importFrom colorspace rainbow_hcl
#' @export
generate_group_colors <- function(n_group, randomize = NULL){
  end_color <- 360 * (n_group - 1) / n_group
  group_color <- rainbow_hcl(n_group, c = 100, start = 0, end = end_color)
  
  
  
  if ((n_group < 5) && is.null(randomize)) {
    do_sample <- FALSE
  } else if ((n_group < 5) && is.logical(randomize)) {
    do_sample <- randomize
  } else if ((n_group >= 5) && is.null(randomize)) {
    do_sample <- TRUE
  } else if ((n_group >= 5) && is.logical(randomize)) {
    do_sample <- randomize
  }
  
  if (do_sample) {
    group_color <- sample(group_color)
  }
  group_color
}

#' cluster and reorder
#' 
#' given a matrix (maybe a distance matrix), cluster and then re-order using
#' dendsort.
#' 
#' @param similarity_matrix matrix of similarities
#' @param matrix_indices indices to reorder
#' @param transform should a transformation be applied to the data first
#' 
#' @import dendsort
#' @importFrom stats as.dist hclust as.dendrogram
#' @export
#' 
#' @return a \link{dendrogram} object. To get the order use \code{order.dendogram}.
similarity_reorder <- function(similarity_matrix, matrix_indices=NULL, transform = "none"){
  if (is.null(matrix_indices)){
    matrix_indices <- seq(1, nrow(similarity_matrix))
  }
  
  if (class(similarity_matrix) != "dist"){
    similarity_matrix <- as.dist(similarity_matrix)
  }
  
  transform_data <- switch(transform,
                           none = similarity_matrix,
                           inverse = 1 / similarity_matrix,
                           sub_1 = 1 - similarity_matrix,
                           log = log(similarity_matrix))
  
  if (min(transform_data) < 0){
    transform_data <- transform_data - min(transform_data)
  }
  
  tmp_clust <- as.dendrogram(hclust(transform_data))
  new_sort <- dendsort(tmp_clust)
  return(list(dendrogram = new_sort, indices = matrix_indices[order.dendrogram(new_sort)]))
  #matrix_indices[order.dendrogram(new_sort)]
}


#' reorder by sample class
#' 
#' to avoid spurious visualization problems, it is useful in a heatmap visualization
#' to reorder the samples within each sample class. This function uses 
#' hierarchical clustering and \link{dendsort} to sort entries in a distance matrix.
#' 
#' @param similarity_matrix matrix of similarities between objects
#' @param sample_classes data.frame or factor denoting classes
#' @param transform a transformation to apply to the data
#' 
#' @details 
#' 
#' The \code{similarity_matrix} should be either a square matrix of similarity values
#' or a distance matrix of class \code{dist}. If your matrix does not encode a "true"
#' distance, you can use a \code{transform} to turn it into a true \code{distance}
#' (for example, if you have correlation, then a distance would be 1 - correlation,
#' use "sub_1" as the transform argument).
#' 
#' The \code{sample_classes} should be either a data.frame or factor argument. If 
#' a data.frame is passed, all columns of the data.frame will be pasted together
#' to create a factor for splitting the data into groups. If the rownames of the
#' data.frame do not correspond to the rownames or colnames of the matrix, then
#' it is assumed that the ordering in the matrix and the data.frame are identical.
#' 
#' @import dendsort
#' @export
#' 
#' @return a list containing the reordering of the matrix in a:
#'   \enumerate{
#'     \item dendrogram
#'     \item numeric vector
#'     \item character vector (will be NULL if rownames are not set on the matrix)
#'   }
#' 
#' @examples 
#' library(visualizationQualityControl)
#' set.seed(1234)
#' mat <- matrix(rnorm(100, 2, sd = 0.5), 10, 10)
#' rownames(mat) <- colnames(mat) <- letters[1:10]
#' neworder <- similarity_reorderbyclass(mat)
#' mat[neworder$indices, neworder$indices]
#'
#' sample_class <- data.frame(grp = rep(c("grp1", "grp2"), each = 5))
#' rownames(sample_class) <- rownames(mat)
#' neworder2 <- similarity_reorderbyclass(mat, sample_class[, "grp", drop = FALSE])
#' 
#' mat[neworder2$indices, neworder2$indices]
#' cbind(neworder$names, neworder2$names)
#' 
similarity_reorderbyclass <- function(similarity_matrix, sample_classes=NULL, transform="none"){
  num_indices <- seq(1, nrow(similarity_matrix))
  
  stopifnot(identical(rownames(similarity_matrix), colnames(similarity_matrix)))
  
  matrix_names <- rownames(similarity_matrix)
  
  if (is.null(sample_classes)){
    sample_classes <- data.frame(none = rep(1, length(num_indices)))
    
    if (!is.null(matrix_names)){
      rownames(sample_classes) <- matrix_names
    }
  }
  
  # transform factor to data.frame because easier to work with one
  # structure later on. Notice that we bet that names of a factor will
  # become the rownames of the data.frame
  if (is.factor(sample_classes)){
    sample_classes <- data.frame(none = sample_classes)
  }
  
  # if the rownames are simply 1-x, then change them to match
  # the matrix rownames so that our checks later work right
  if (identical(rownames(sample_classes), as.character(seq(1, nrow(similarity_matrix))))){
    rownames(sample_classes) <- rownames(similarity_matrix)
  }
  
  class_names <- rownames(sample_classes)
  # check if the names are the same, and in the same order, because we are not
  # going to attempt anything otherwise, because all of the heatmap stuff (that
  # we mostly use this for) depend on it. Note that we set the rownames and colnames
  # on the matrix if they are missing so the dendrogram has something useful
  # on it!
  if (!is.null(matrix_names)){
    if (!identical(matrix_names, class_names)){
      stop("similarity_matrix and sample_classes rownames must match!", call. = FALSE)
    }
  } else {
    rownames(similarity_matrix) <- colnames(similarity_matrix) <- num_indices
  }
  
  # we use ALL of the columns to generate factors that are used for splitting!
  if (ncol(sample_classes) > 1) {
    use_class <- do.call(paste, c(sample_classes, sep="."))
  } else {
    use_class <- sample_classes
  }
  
  
  split_indices <- split(num_indices, use_class)
  
  new_order <- lapply(split_indices, function(x){
    similarity_reorder(similarity_matrix[x, x], x, transform = transform)
  })
  
  out_dendrogram <- new_order[[1]][["dendrogram"]]
  
  if (length(new_order) > 1){
    for (id in seq(2, length(new_order))){
      out_dendrogram <- merge(out_dendrogram, new_order[[id]][["dendrogram"]], adjust = "add.max")
    }
  }
  
  out_indices <- unlist(lapply(new_order, function(x){x$indices}), use.names = FALSE)
  out_names <- NULL
  
  if (!is.null(matrix_names)){
    out_names <- matrix_names[out_indices]
  }
  
  return(list(dendrogram = out_dendrogram, indices = out_indices, names = out_names))
}


#' easier heatmaps
#' 
#' rolls some of the common \code{Heatmap} options into a single function call
#' to make life easier when creating lots of heatmaps. \strong{Note:} clustering
#' of rows and columns is disabled, it is expected that you are reordering the
#' matrix beforehand, or passing in \code{column_order} and \code{row_order} as
#' arguments to be passed to \code{Heatmap} (see example). Matrices can be reordered
#' using \code{\link{similarity_reorderbyclass}}, and nice class colors generated
#' using \code{\link{generate_group_colors}}
#' 
#' @param matrix_data the matrix you want to plot as a heatmap
#' @param color_values the color mapping of values to colors (see Details)
#' @param title what do the values represent
#' @param row_color_data data for row annotations
#' @param row_color_list list for row annotations
#' @param col_color_data data for column annotations
#' @param col_color_list list for column annotations
#' @param ... other \code{Heatmap} parameters
#' 
#' @details This function uses the \code{ComplexHeatmap} package to produce
#' heatmaps with complex row- and column-color annotations. Both \code{row_color_data}
#' and \code{col_color_data} should be \code{data.frame}'s where each column describes
#' meta-data about the rows or columns of the matrix. The \code{row_color_list} and 
#' \code{col_color_list} provide the mapping of color to annotation, where each
#' \code{list} entry should be a named vector of colors, with the list entry
#' corresponding to a column entry in the data.frame, and the names of the colors
#' corresponding to annotations in that column.
#' 
#' @examples  
#' \dontrun{
#' library(circlize)
#' data(grp_cor)
#' data(grp_info)
#' colormap <- colorRamp2(c(0, 1), c("black", "white"))
#' 
#' annotation_color <- c(grp1 = "green", grp2 = "red", set1 = "blue",
#'                       set2 = "yellow")
#' 
#' row_data <- grp_info[, "grp", drop = FALSE]
#' col_data <- grp_info[, "set", drop = FALSE]
#' row_annotation = list(grp = annotation_color[1:2])
#' col_annotation = list(set = annotation_color[3:4])
#' 
#' visqc_heatmap(grp_cor, colormap, row_color_data = row_data, row_color_list = row_annotation,
#'                  col_color_data = col_data, col_color_list = col_annotation)
#'                  
#' reorder_sim <- similarity_reorderbyclass(grp_cor, transform = "sub_1")
#' visqc_heatmap(grp_cor, colormap, "reorder1", row_data, row_annotation, col_data, col_annotation,
#'                  column_order = reorder_sim$indices, row_order = reorder_sim$indices)
#' 
#' sample_classes <- grp_info[, "grp", drop = FALSE]
#' reorder_sim2 <- similarity_reorderbyclass(grp_cor, sample_classes, "sub_1")
#' visqc_heatmap(grp_cor, colormap, "reorder2", row_data, row_annotation, col_data, col_annotation,
#'                  column_order = reorder_sim2$indices, row_order = reorder_sim2$indices)
#' }
#' 
#' @import ComplexHeatmap
#' @export
visqc_heatmap <- function(matrix_data, color_values, title = "", row_color_data = NULL, row_color_list = NULL, col_color_data = NULL, col_color_list = NULL, ...){
  if (!is.null(row_color_data) && !is.null(row_color_list)){
    row_annot <- rowAnnotation(df = row_color_data, col = row_color_list)
  } else{
    row_annot = new("HeatmapAnnotation")
  }
  if (!is.null(col_color_data) && !is.null(col_color_list)){
    col_annot <- HeatmapAnnotation(df = col_color_data, col = col_color_list)
  } else{
    col_annot <- new("HeatmapAnnotation")
  }
  
  heat_out <- Heatmap(matrix_data, col = color_values,
                      top_annotation = col_annot, column_title = title, 
                      cluster_rows = FALSE, cluster_columns = FALSE, ...) + row_annot
  heat_out
}
